### PR TITLE
[MXNET-139] Tutorial for mixed precision training with float16

### PR DESCRIPTION
**Pull request by rahul003**
## Description ##
Adds a FAQ page on mixed precision training with float16. Explains usage for both Gluon and Symbolic, and discusses tips to improve performance and accuracy when using mixed precision. 

https://issues.apache.org/jira/browse/MXNET-139

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Adds tutorial
- [x] Updates symbolic examples for training and fine-tuning
- [x] Updates Gluon example

### CODE REVIEW

File: docs/tutorials/python/float16.md [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+- CUDNN v7 or higher
+
+## Using the Gluon API
+
+With Gluon, we need to take care of two things to convert a model to support float16.
+1. Cast the Gluon Block, so as to cast the parameters of layers and change the type of input expected, to float16.
+2. Cast the data to float16 to match the input type expected by the blocks if necessary.
+
+### Training
+Let us look at an example of training a Resnet50 model with the Caltech101 dataset with float16.
```
Comment (eric-haibin-lin) [inline docs/tutorials/python/float16.md]: Add a reference link to the dataset description?

File: docs/tutorials/python/float16.md [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+import time
+import numpy as np
+import mxnet as mx
+from mxnet import nd, autograd, gluon
+from mxnet.gluon.model_zoo import vision as models
+from mxnet.metric import Accuracy
+from mxnet.gluon.data.vision.datasets import ImageFolderDataset
+```
+
+Let us start by fetching the Caltech101 dataset and extracting it.
```
Comment (eric-haibin-lin) [inline docs/tutorials/python/float16.md]: Could you add a reminder of how big the dataset is (num images, number of GBs)

File: docs/tutorials/python/float16.md [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+The Volta range of Graphics Processing Units (GPUs) from Nvidia have Tensor Cores which perform efficient float16 computation. A tensor core allows accumulation of half precision products into single or half precision outputs. For the rest of this tutorial we assume that we are working with Nvidia's Tensor Cores on a Volta GPU.
+
+In this tutorial we will walk through how one can train deep learning neural networks with mixed precision on supported hardware. We will first see how to use float16 and then some techniques on achieving good performance and accuracy.
+
+## Prerequisites
+
+- Volta range of Nvidia GPUs
+- Cuda 9 or higher
+- CUDNN v7 or higher
+
```
Comment (eric-haibin-lin) [inline docs/tutorials/python/float16.md]: Could you start with an overview that the tutorial covers both Gluon and Symbolic APIs?

File: docs/tutorials/python/float16.md [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+# Creating the network
+def get_network(dtype):
+    net = models.get_model(name='resnet50_v2', ctx=ctx, pretrained=False, classes=101)
+    net.collect_params().initialize(mx.init.Xavier(magnitude=2.24), ctx=ctx)
+    net.hybridize()
+    net.cast(dtype)
+    return net
+```
+
+It is preferable to use **multi_precision mode of optimizer** when training in float16. This mode of optimizer maintains the weights in float32 even when the training is in float16. This helps increase precision of the weights and leads to faster convergence for some networks. (Further discussion on this towards the end.)
```
Comment (eric-haibin-lin) [inline docs/tutorials/python/float16.md]: Do all optimizers support this mode?

**Comment by mli** [comment]
I feel this tutorial is a little bit too complex to reader to follow.

1. show how to write a simple dense and conv layer to use fp16, show performance numbers
2. show an end-to-end example based on another tutorial. such as http://gluon-crash-course.mxnet.io/ (it's not a good exmaple because too simplified), mentioned to users that read it first. and this tutorial will only highlight the difference for using fp16. 
3. show performance numbers and training curves

File: docs/tutorials/python/float16.md [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+
+The float16 data type, is a 16 bit floating point representation according to the IEEE 754 standard. It has a dynamic range where the precision can go from 0.0000000596046 (highest, for values closest to 0) to 32 (lowest, for values in the range 32768-65536). Despite the decreased precision when compared to single precision (float32), float16 computation can be much faster on supported hardware. The motivation for using float16 for deep learning comes from the idea that deep neural network architectures have natural resilience to errors due to backpropagation. Half precision is typically sufficient for training neural networks. This means that on hardware with specialized support for float16 computation we can greatly improve the speed of training and inference. This speedup results from faster matrix multiplication, saving on memory bandwidth and reduced communication costs. It also reduces the size of the model, allowing us to train larger models and use larger batch sizes.
```
Comment (aaronmarkham) [inline docs/tutorials/python/float16.md]: no comma after type

File: docs/tutorials/python/float16.md [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+
+The Volta range of Graphics Processing Units (GPUs) from Nvidia have Tensor Cores which perform efficient float16 computation. A tensor core allows accumulation of half precision products into single or half precision outputs. For the rest of this tutorial we assume that we are working with Nvidia's Tensor Cores on a Volta GPU.
+
+In this tutorial we will walk through how one can train deep learning neural networks with mixed precision on supported hardware. We will first see how to use float16 and then some techniques on achieving good performance and accuracy.
```
Comment (aaronmarkham) [inline docs/tutorials/python/float16.md]: I'd move this to the top as the main intro, then use `## Background` for the rest.

Comment (aaronmarkham) [inline docs/tutorials/python/float16.md]: In this tutorial you will learn how you can train...
You will first see how...

Please continue on in this pattern.

File: docs/tutorials/python/float16.md [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+    return transposed, label
+
+dataset_train = ImageFolderDataset(root=training_path, transform=transform)
+dataset_test = ImageFolderDataset(root=testing_path, transform=transform)
+
+train_data = gluon.data.DataLoader(dataset_train, BATCH_SIZE, shuffle=True, num_workers=NUM_WORKERS)
+test_data = gluon.data.DataLoader(dataset_test, BATCH_SIZE, shuffle=False, num_workers=NUM_WORKERS)
+```
+
+Next, we'll define softmax cross entropy as our loss, accuracy as our metric and the context on which to run our training jobs. It is set by default to gpu. Please note that float16 on CPU might not be supported for all operators, as float16 on CPU is slower than float32.
```
Comment (aaronmarkham) [inline docs/tutorials/python/float16.md]: GPU or `gpu`?

File: docs/tutorials/python/float16.md [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+net = get_network(DTYPE)
+train(net, dtype=DTYPE, num_epochs=25)
+```
+
+Note the accuracy you observe above. You can change DTYPE above to float32 if you want to observe the speedup gained by using float16.
+
+
+### Finetuning
+
+You can also finetune in float16, a model which was originally trained in float32. The section of the code which builds the network would now look as follows. We first fetch the pretrained resnet50_v2 model from model zoo. This was trained using Imagenet data, so we need to pass classes as 1000 for fetching the pretrained model. Then we create our new network for Caltech 101 by passing number of classes as 101. We will then cast it to `float16` so that we cast all parameters to `float16`.
```
Comment (aaronmarkham) [inline docs/tutorials/python/float16.md]: the [model zoo](http://mxnet.incubator.apache.org/api/python/gluon/model_zoo.html).

File: docs/tutorials/python/float16.md [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+
+## Using the Symbolic API
+
+Training a network in float16 with the Symbolic API involves the following steps.
+1. Add a layer at the beginning of the network, to cast the data to float16. This will ensure that all the following layers compute in float16.
+2. It is advisable to cast the output of the layers before softmax to float32, so that the softmax computation is done in float32. This is because softmax involves large reductions and it helps to keep that in float32 for more precise answer.
+
+There are a few examples of building such networks which can handle float16 input in [examples/image-classification/symbols/](https://github.com/apache/incubator-mxnet/tree/master/example/image-classification/symbols). Specifically you could look at the [resnet](https://github.com/apache/incubator-mxnet/blob/master/example/image-classification/symbols/resnet.py) example.
+
+An illustration of the relevant section of the code is below.
```
Comment (aaronmarkham) [inline docs/tutorials/python/float16.md]: Try to avoid above below left and right. Use follows or as follows. Or previously. This supports different reading modes.

File: docs/tutorials/python/float16.md [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+```
+python fine-tune.py --network resnet --num-layers 50 --pretrained-model imagenet1k-resnet-50 --data-train ~/efs/data/caltech-256/caltech256-train.rec --data-val ~/efs/data/caltech-256/caltech256-val.rec --num-examples 15420 --num-classes 256 --gpus 0 --batch-size 64 --dtype float16
+```
+
+## Things to keep in mind
+
+### For performance
+1. Nvidia Tensor core essentially perform the computation D = A * B + C, where A and B are half precision matrices, while C and D could be either half precision or full precision. The tensor cores are most efficient when dimensions of these matrices are multiples of 8. This means that Tensor Cores can not be used in all cases for fast float16 computation. When training models like Resnet50 on the Cifar10 dataset, the tensors involved are sometimes smaller, and tensor cores can not always be used. The computation in that case falls back to slower algorithms and using float16 turns out to be slower than float32 on a single GPU. Note that when using multiple GPUs, using float16 can still be faster than float32 because of reduction in communication costs.
```
Comment (aaronmarkham) [inline docs/tutorials/python/float16.md]: cores perform or core performs

Use consistent CapitaliZation throughout.

**Comment by anirudh2290** [comment]
ping @rahul003

**Comment by rahul003** [comment]
Thanks guys for your comments. I'll address them soon and update the PR

File: docs/tutorials/python/float16.md [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+optimizer = mx.optimizer.create('sgd', multi_precision=True, rescale_grad=1.0/128)
+```
+
+Networks like Multibox SSD, R-CNN, bigLSTM and Seq2seq were found to exhibit such behavior. Refer the linked articles below for more details on this.
+    
+## References
+1. [Training with Mixed Precision User Guide](http://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html)
+2. [Mixed Precision Training at ICLR 2018](https://arxiv.org/pdf/1710.03740.pdf)
+3. [Mixed-Precision Training of Deep Neural Networks](https://devblogs.nvidia.com/mixed-precision-training-deep-neural-networks/)
+
```
Comment (ThomasDelteil) [inline docs/tutorials/python/float16.md]: Can you add this `<!-- INSERT SOURCE DOWNLOAD BUTTONS -->` at the end of your .md file to enable the notebook download of your tutorial? Thanks!

File: docs/tutorials/python/float16.md [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+```
+
+Let us start by fetching the Caltech101 dataset and extracting it. 
+
+
+```python
+url = "https://s3.us-east-2.amazonaws.com/mxnet-public/101_ObjectCategories.tar.gz"
+dataset_name = "101_ObjectCategories"
+data_folder = "data"
+if not os.path.isdir(data_folder):
```
Comment (ThomasDelteil) [inline docs/tutorials/python/float16.md]: these lines are un necessary as the `mx.gluon.utils.download` will create the directory if it does not exist 😃

File: docs/tutorials/python/float16.md [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+
+You can also finetune in float16, a model which was originally trained in float32. The section of the code which builds the network would now look as follows. We first fetch the pretrained resnet50_v2 model from model zoo. This was trained using Imagenet data, so we need to pass classes as 1000 for fetching the pretrained model. Then we create our new network for Caltech 101 by passing number of classes as 101. We will then cast it to `float16` so that we cast all parameters to `float16`.
+
+
+```python
+def get_pretrained_net(dtype):
+    pretrained_net = models.get_model(name='resnet50_v2', ctx=ctx, pretrained=True, classes=1000)
+    pretrained_net.hybridize()
+    pretrained_net.cast(dtype)
+
```
Comment (ThomasDelteil) [inline docs/tutorials/python/float16.md]: a simpler way of fine-tuning a model from the model-zoo is to use `pretrained_net.output = gluon.nn.Dense(101)` and then initializing it.

**Comment by piiswrong** [comment]
@rahul003 ping

**Comment by ThomasDelteil** [comment]
@rahul003 ping, do you plan to finish this tutorial? It is dearly needed.

File: docs/tutorials/python/float16.md (hunk 4 above)
Comment (rahul003) [reply docs/tutorials/python/float16.md]: SGD supports this natively, as in there's a special kernel for that. Other optimizers support this by making a copy in Python, which can be slightly slower.

**Comment by rahul003** [comment]
On second thoughts as per feedback above I changed this from a runnable tutorial style to a document focusing on the changes needed to switch to Mixed precision. I updated an example in the source to add the example I had in this tutorial, and put a command to run that in this document. 

I've setup two runs whose plots I will add tomorrow. 

I'm hesitant to share raw performance numbers as those would soon become outdated as we improve. I could mention a rough speedup factor instead. What do you guys think?

**Comment by rahul003** [comment]
@reviewers, please check the tutorial now. I think we can merge it and keep updating it if you have suggestions for other things to add. Even as is, it would be very useful.

**Comment by rahul003** [comment]
can we merge this?

File: example/image-classification/benchmark_score.py [hunk 14]
```diff
@@ -74,11 +75,17 @@ def score(network, dev, batch_size, num_batches):
...
         for d in devs:
             logging.info('device: %s', d)
             for b in batch_sizes:
-                speed = score(network=net, dev=d, batch_size=b, num_batches=10)
-                logging.info('batch size %2d, image/sec: %f', b, speed)
+                for dtype in ['float32', 'float16']:
+                    if d == mx.cpu() and dtype == 'float16':
+                        #float16 is not supported on CPU
+                        continue
+                    elif net in ['inception-bn', 'alexnet'] and dt == 'float16':
```
Comment (xinyu-intel) [inline example/image-classification/benchmark_score.py]: I guess this `dt` here doesn't work, please check again. By the way, it's necessary to add a logging here to notice user what the dtype is.

Comment (rahul003) [reply example/image-classification/benchmark_score.py]: Thanks for letting me know, fixing it here https://github.com/apache/incubator-mxnet/pull/11533/
